http-auth/routes/routes_test.go

```diff
@@ -0,0 +1,179 @@
+package routes
+
+import (
+	"html"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"golang.org/x/time/rate"
+)
+
+func TestMain(t *testing.T) {
+	t.Run("GET /", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/", nil)
+		rr := httptest.NewRecorder()
+
+		IndexHandler(rr, request)
+
+		AssertStatusOK(t, rr)
+	})
+
+	t.Run("POST /", func(t *testing.T) {
+		bodyHTML := "<em>Hello Test</em>"
+		b := strings.NewReader(bodyHTML)
+		request, _ := http.NewRequest(http.MethodPost, "/", b)
+		request.Header.Add("Content-Type", "text/html")
+
+		rr := httptest.NewRecorder()
+
+		IndexHandler(rr, request)
+
+		AssertStatusOK(t, rr)
+
+		AssertBodyEquals(t, rr, HTMLHeader+html.EscapeString(bodyHTML))
+	})
+
+	t.Run("GET /200", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/200", nil)
+		rr := httptest.NewRecorder()
+
+		Handle200(rr, request)
+
+		AssertStatusOK(t, rr)
+	})
+
+	t.Run("GET /500", func(t *testing.T) {
+		request, _ := http.NewRequest(http.MethodGet, "/500", nil)
+		rr := httptest.NewRecorder()
+
+		Handle500(rr, request)
+
+		AssertStatusInternalServerError(t, rr)
+	})
+
+	t.Run("GET /authenticated", func(t *testing.T) {
+		t.Run("no authorisation provided", func(t *testing.T) {
+
+			request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+			rr := httptest.NewRecorder()
+
+			handler := HandleAuthenticated("test", "test")
+
+			handler(rr, request)
+
+			AssertStatusUnauthorized(t, rr)
+		})
+
+		t.Run("invalid authorisation provided", func(t *testing.T) {
+
+			request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+			request.Header.Add("Authorization", "Basic invalid")
+			rr := httptest.NewRecorder()
+
+			handler := HandleAuthenticated("test", "test")
+
+			handler(rr, request)
+
+			AssertStatusUnauthorized(t, rr)
+		})
+
+		t.Run("valid authorisation provided", func(t *testing.T) {
+
+			request, _ := http.NewRequest(http.MethodGet, "/authenticated", nil)
+			request.Header.Add("Authorization", "Basic dGVzdHVzZXI6c29tZXN0cm9uZ1BXRA==")
```

Comment:
I'd probably add a comment here explaining how this string was generated, so that someone can more easily verify it in the future, and particularly if they need to change the username/password for the test, so they know how to.

Reply:
I have updated it to use `SetBasicAuth` method, which makes it easier to understand. Thanks!